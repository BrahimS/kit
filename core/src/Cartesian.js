import React, { Fragment } from 'react'

import { cartesianProduct } from './util'

export default ({ component, ...props }) => {
  const combinations = cartesianProduct(props)
  const Component = component

  return (
    <div>
      {combinations.map((props, i) =>
        <Component key={i} {...props} />
      )}
    </div>
  )
}
