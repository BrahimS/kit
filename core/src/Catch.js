import React from 'react'

class Catch extends React.Component {
  constructor() {
    super()

    this.state = {}
  }

  componentDidCatch(err) {
    this.setState({ err })
  }

  componentWillReceiveProps(next) {
    if (next.children !== this.props.children) {
      this.setState({ err: null })
    }
  }

  render() {
    const { err } = this.state

    if (err) return <pre>{err.toString()}</pre>

    return this.props.children
  }
}

export default Catch
