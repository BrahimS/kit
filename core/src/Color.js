import React from 'react'
import contrast from 'get-contrast'
import { Debug } from '.'
import { flatten } from './util'
import { Box, Code, Flex, Text } from './ui'

export const Swatch = ({ name, value, ...props }) => (
  <React.Fragment>
    <Box p={3} {...props}>
      <Box bg={value} p={[3, 4, 5]} />
      <Text>{name}</Text>
      <Code f={0}>{value}</Code>
    </Box>
  </React.Fragment>
)

export const Group = ({ colors, prefix, ...props }) => (
  <Flex {...props}>
    {colors.map((c, i) => {
      let name = null
      let value = null
      if (typeof c === 'object') {
        name = Object.keys(c)[0]
        value = c[name]
      } else {
        name = prefix ? `${prefix}.${i}` : c
        value = c
      }

      const color = contrast.isAccessible(value, 'white') ? 'white' : 'black'

      return (
        <Box key={name} bg={value} color={color} p={3}>
          <Text >{name}</Text>
        </Box>
      )
    })}
  </Flex>
)

export const Contrast = ({ color, bg, children }) => (
  <Box p={[3, 4, 5]} bg={bg} color={color}>
    <Text f={[5, 6, 128]} fontWeight='bold'>
      Aa
    </Text>
    <Flex alignItems='baseline'>
      <Code f={6}>{contrast.ratio(bg, color).toFixed(2)}</Code>
      <Text f={4} fontWeight='bold'>{contrast.score(bg, color)}</Text>
    </Flex>
    {children}
  </Box>
)

const Color = ({ colors }) => {
  const colorsObj = flatten(colors)
  const colorsArr = Object.keys(colorsObj).reduce((acc, key) =>
    acc.concat([{ key, name: key, value: colorsObj[key] }])
  , [])

  return (
    <Flex
      flexWrap='wrap'
      children={colorsArr.map(color =>
        <Swatch
          width={[1/2, 1/3, 1/5]}
          {...color}
        />
      )}
    />
  )
}

Color.Group = Group
Color.Swatch = Swatch
Color.Contrast = Contrast

export default Color
