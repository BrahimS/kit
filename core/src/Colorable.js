import React from 'react'
import colorable from 'colorable'
import { flatten } from './util'

export default ({ colors, children, ...opts }) => {
  const allColors = Array.isArray(colors) ? colors : flatten(colors)

  return children(colorable(allColors, opts))
}
