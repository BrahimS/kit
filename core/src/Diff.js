import React, { Fragment } from 'react'
import nano from 'nano-style'
import { Box } from './ui'

const DiffBox = nano(Box)(({ blendMode }) => ({
  position: 'relative',
  mixBlendMode: blendMode || 'multiply'
}))

const ComponentBox = nano(Box)(({ blendMode }) => ({
  position: 'absolute',
  mixBlendMode: blendMode || 'multiply'
}))

export default ({ blendMode, a, b }) => (
  <DiffBox>
    <ComponentBox blendMode={blendMode}>{a}</ComponentBox>
    <ComponentBox blendMode={blendMode}>{b}</ComponentBox>
  </DiffBox>
)
