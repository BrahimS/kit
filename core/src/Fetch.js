import React, { Component } from 'react'
import fetch from 'isomorphic-fetch'

const fetchStates = ['loading', 'fetched', 'error']

export default class Fetch extends Component {
  constructor() {
    super()

    this.state = {
      fetchState: 'loading',
      data: null,
      error: null,
      loading: true,
      fetching: true,
      fetched: false
    }
  }

  async componentDidMount() {
    this.fetchData()
  }

  fetchData = async () => {
    const { url } = this.props

    try {
      const res = await fetch(url)
      const data = await res.json()

      this.setState({
        fetchState: 'fetched',
        editedData: data,
        data
      })
    } catch (error) {
      this.setState({
        error,
        fetchState: 'error'
      })
    }

    this.setState({
      loading: false,
      fetching: false,
      fetched: true
    })
  }

  handleRefetchClick = () => this.fetchData()

  handleDataChange = value => {
    if (this.state.fetching) {
      return
    }

    try {
      const data = JSON.parse(value)
      this.setState({ data, editedData: value })
    } catch (e) {
      this.setState({ editedData: value })
    }
  }

  handleFetchStateChange = e => {
    if (this.state.fetching) {
      return
    }

    const { value } = e.target

    this.setState({
      fetchState: value,
      loading: value === 'loading',
      error: value === 'error'
    })
  }

  render = () => this.props.children({
    fetchStates,
    onRefetchClick: this.handleRefetchClick,
    onDataChange: this.handleDataChange,
    onFetchStateChange: this.handleFetchStateChange,
    ...this.state
  })
}
