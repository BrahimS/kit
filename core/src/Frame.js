import React from 'react'
import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'
import { renderToStaticMarkup } from 'react-dom/server'
import nano from 'nano-style'

class Frame extends React.Component {
  static propTypes = {
    head: PropTypes.node,
    zoom: PropTypes.number,
    width: PropTypes.string,
    height: PropTypes.string,
    css: PropTypes.string,
  }

  static defaultProps = {
    zoom: 1,
    width: '100%',
    height: '100%',
    css: 'body{font-family:system-ui,sans-serif;line-height:1.5}'
  }

  constructor() {
    super()

    this.doc = null
    this.win = null
    this.div = null

    this.getSrc = () => {
      const { zoom = 1, css = '', head } = this.props
      let headHTML = ''
      if (head) {
        headHTML = renderToStaticMarkup(head)
      }
      return `<style>*{box-sizing:border-box}body{margin:0;min-height:100vh;zoom:${zoom}} ${css}</style>${headHTML}
<div id='app'></div>`
    }

    this.onLoad = e => {
      this.doc = this.root.contentDocument
      this.win = this.root.contentWindow
      this.update(this.props)
    }

    this.update = ({ render, children }) => {
      if (!this.doc) return
      const div = this.doc.getElementById('app')
      if (typeof render === 'function') {
        ReactDOM.render(
          render({
            document: this.doc,
            window: this.win
          }),
          div
        )
      } else {
        ReactDOM.render(children, div)
      }
    }
  }

  componentWillReceiveProps(next) {
    if (next.children !== this.props.children) {
      this.update(next)
    }
  }

  render() {
    const { width, height, zoom, children } = this.props

    return (
      <iframe
        ref={ref => (this.root = ref)}
        style={{
          width,
          height,
          zoom,
          pointerEvents: 'none',
          display: 'block',
          margin: 0,
          overflow: 'scroll',
          backgroundColor: '#fff',
          opacity: !!children ? 1 : 0.25,
          border: 0
        }}
        srcDoc={this.getSrc()}
        scrolling="yes"
        onLoad={this.onLoad}
      />
    )
  }
}

export default Frame
