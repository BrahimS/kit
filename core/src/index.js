export {
  Library,
  Example,
  Detail
} from './Library'
export { default as LiveEditor } from './LiveEditor'
export { default as Frame } from './Frame'
export { default as Catch } from './Catch'
export { default as XRay } from './XRay'
export { default as PropsForm } from './PropsForm'
export { default as Responsive } from './Responsive'
export { default as Cartesian } from './Cartesian'
export { default as Matrix } from './Matrix'
export { default as Diff } from './Diff'
export { default as Debug } from './Debug'
export { default as TypeScale } from './TypeScale'
export { default as Style } from './Style'
export { default as Font } from './Font'
export { default as UI } from './ui'
