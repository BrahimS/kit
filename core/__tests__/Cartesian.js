import React from 'react'
import styled from 'styled-components'
import { render } from 'react-testing-library'
import { color, space, fontSize } from 'styled-system'

import { Cartesian } from '../src'

const buttonProps = {
  children: 'Hello, world!',
  fontSize: [1, 2, 3],
  px: [2, 3],
  backgroundColor: ['pink', 'tomato', 'purple']
}

const Button = styled.a`
  ${color}
  ${fontSize}
  ${space}
`

test('Cartesian renders all examples', t => {
  const { container } = render(<Cartesian {...buttonProps} component={Button} />)

  expect(container).toMatchSnapshot()
})
