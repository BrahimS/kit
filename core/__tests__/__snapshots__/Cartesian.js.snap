// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cartesian renders all examples 1`] = `
<div>
  <a
    class="sc-kvZOFW oRYKK"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW oRYKK"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW oRYKK"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW fBqLEr"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW fBqLEr"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW fBqLEr"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW hDLVbT"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW hDLVbT"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW hDLVbT"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW isfRCQ"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW isfRCQ"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW isfRCQ"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW dqCqkj"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW dqCqkj"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW dqCqkj"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW ergpZW"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW ergpZW"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-kvZOFW ergpZW"
    font-size="3"
  >
    Hello, world!
  </a>
</div>
`;
